include/deal.II/physics/relations.h

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright (C) 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_relations_h
+#define dealii_relations_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/tensor.h>
+
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Physics
+{
+  /**
+   * Functions to determine relations between vectors.
+   */
+  namespace Relations
+  {
+    /**
+     * Calculate the angle $\theta$ between two vectors @p a and @p b. The returned
+     * angle will be in the range $[0, \pi]$.
+     *
+     * This function uses the geometric definition of the scalar product.
+     * @f[
+     *   \vec{a} \cdot \vec{b} = \|\vec{a}\| \|\vec{b}\| \cos(\theta)
+     * @f]
+     */
+    template <int spacedim, typename Number>
+    Number
+    angle(const Tensor<1, spacedim, Number> &a,
+          const Tensor<1, spacedim, Number> &b);
+
+    /**
+     * Calculate the angle $\theta$ between two vectors @p a and @p b, where both
+     * vectors are located in a plane described by a normal vector @p axis. The
+     * angle describes the relation with respect to the vector @p axis and will
+     * be in the range $[-\pi, \pi]$.
+     *
+     * The vector @p axis needs to be a unit vector and perpendicular to both
+     * vectors @p a and @p b.
+     *
+     * This function uses the geometric definitions of both the scalar and cross
+     * product.
+     * @f{align*}{
+     *   \vec{a} \cdot  \vec{b} &= \|\vec{a}\| \|\vec{b}\| \cos(\theta) \\
+     *   \vec{a} \times \vec{b} &= \|\vec{a}\| \|\vec{b}\| \sin(\theta) \vec{n}
+     * @f}
+     * We can create the tangent of the angle using both products.
+     * @f[
+     *   \tan{\theta}
+     *   = \frac{\sin(\theta)}{\cos(theta)}
+     *   = \frac{(\vec{a} \times \vec{b}) \cdot \vec{n}}{\vec{a} \cdot \vec{b}}
+     * @f]
+     *
+     * @note Only applicable for three-dimensional vectors `spacedim == 3`.
+     */
+    template <int spacedim, typename Number>
+    Number
+    angle(const Tensor<1, spacedim, Number> &a,
+          const Tensor<1, spacedim, Number> &b,
+          const Tensor<1, spacedim, Number> &axis);
+  } // namespace Relations
+} // namespace Physics
+
+
+
+#ifndef DOXYGEN
+
+
+
+template <int spacedim, typename Number>
+inline Number
+Physics::Relations::angle(const Tensor<1, spacedim, Number> &a,
+                          const Tensor<1, spacedim, Number> &b)
+{
+  Assert(a.norm() > 1.e-12 && b.norm() > 1.e-12,
```

Comment:
I'm not generally a fan of these sorts of comparisons because the vectors may well have physical units -- and in that case, what is small and what is large is all relative. I wouldn't object to a comparison of the form
```suggestion
  Assert(a.norm() > 1.e-12*b.norm() && b.norm() > 1.e-12*a.norm(),
```